### PR TITLE
Integrate new marketplace

### PR DESCRIPTION
**Pull request by DeepDiver1975**

### CODE REVIEW

File: lib/AppStore/Service.php [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+use OCP\IAppConfig;
+
+class Service {
+
+	public function __construct(IAppManager $appManager, IAppConfig $appConfig) {
+		$this->appManager = $appManager;
+		$this->appConfig = $appConfig;
+	}
+
+	public function installApp($ocsId) {
```
Comment (PVince81) [inline lib/AppStore/Service.php]: some PHPDoc would be nice, at least for the public methods

File: lib/AppStore/Service.php [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		}
+
+		// download package
+		$package = $this->downloadPackage($ocsId);
+		$this->appManager->installApp($package);
+	}
+
+	private function downloadPackage($ocsId) {
+		// implementation based on the current appstore - needs factory
+		$ocsClient = new OCSClient(
```
Comment (PVince81) [inline lib/AppStore/Service.php]: No OCSClientFactory or something for better DI ? 😉

File: lib/AppStore/Service.php [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+		$download = $ocsClient->getApplicationDownload($ocsId, \OCP\Util::getVersion());
+		if(isset($download['downloadlink']) and $download['downloadlink'] !== '') {
+			// Replace spaces in download link without encoding entire URL
+			$downloadLink = str_replace(' ', '%20', $download['downloadlink']);
+
+			$pathInfo = pathinfo($downloadLink);
+			$extension = isset($pathInfo['extension']) ? '.' . $pathInfo['extension'] : '';
+			$path = \OC::$server->getTempManager()->getTemporaryFile($extension);
+			$client = \OC::$server->getHTTPClientService()->newClient();
+			$client->get($downloadLink, ['save_to' => $path]);
```
Comment (PVince81) [inline lib/AppStore/Service.php]: Will this throw an exception if the download failed ?

File: lib/AppStore/Service.php [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+		throw new \Exception('Unknown app');
+	}
+
+	public function isAppInstalled($ocsId) {
+		$info = $this->getInstalledAppInfo($ocsId);
+		return !is_null($info);
+	}
+
+	public function updateAvailable($ocsId) {
```
Comment (PVince81) [inline lib/AppStore/Service.php]: since this returns a version number, call it `getAvailableUpdateVersion()`

File: lib/AppStore/Service.php [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	private function getAppInfo($ocsId) {
+		//
+		// TODO: read from cached information as stored by using ./occ market:update
+		//
+		$ocsClient = new OCSClient(
+			\OC::$server->getHTTPClientService(),
+			\OC::$server->getConfig(),
+			\OC::$server->getLogger()
+		);
+		return $ocsClient->getApplication($ocsId, \OCP\Util::getVersion());
```
Comment (PVince81) [inline lib/AppStore/Service.php]: error/exception handling ?

File: lib/AppStore/Service.php [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+			$info = $this->appManager->getAppInfo($app);
+			if (isset($info['ocsid']) && (int)$info['ocsid'] === (int)$ocsId) {
+				return $info;
+			}
+		}
+
+		return null;
+	}
+
+	public function updateApp($ocsId) {
```
Comment (PVince81) [inline lib/AppStore/Service.php]: add PHPDoc, from what I see this downloads the app, extracts it then installs it over the old version

File: lib/AppStore/Service.php [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+					$ocsId = $info['ocsid'];
+					$newVersion = $this->updateAvailable($ocsId);
+					if ($newVersion) {
+						$result[$app] = [
+							'version' => $newVersion,
+							'ocsid' => $ocsId
+						];
+					}
+				} catch (\InvalidArgumentException $ex) {
+				}
```
Comment (PVince81) [inline lib/AppStore/Service.php]: add PHPDoc that explains why ignoring this exception is intentional

File: lib/MarketService.php [hunk 8]
```diff
@@ -22,130 +22,40 @@
...
-	}
-
-	public function installApp($ocsId) {
+	public function __construct(IConfig $config, IAppManager $appManager, IAppConfig $appConfig) {
+		$storeUrl = $config->getSystemValue('appstoreurl', 'https://api.owncloud.com/v1');
 
-		$info = $this->getInstalledAppInfo($ocsId);
-		if (!is_null($info)) {
-			throw new \InvalidArgumentException('App is already installed');
+		if ($storeUrl === 'https://api.owncloud.com/v1') {
```
Comment (PVince81) [inline lib/MarketService.php]: Is that a fallback for the old app store ? If yes, add a comment that says so

File: lib/MarketService.php [hunk 9]
```diff
@@ -22,130 +22,40 @@
...
-			$path = \OC::$server->getTempManager()->getTemporaryFile($extension);
-			$client = \OC::$server->getHTTPClientService()->newClient();
-			$client->get($downloadLink, ['save_to' => $path]);
-
-			return $path;
-		}
-
-		throw new \Exception('Unknown app');
+	public function installApp($ocsId) {
+		$this->impl->installApp($ocsId);
```
Comment (PVince81) [inline lib/MarketService.php]: return code / thrown exceptions ?

File: lib/Marketplace/Service.php [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+ *
+ */
+
+
+namespace OCA\Market\Marketplace;
+
+use OCP\App\IAppManager;
+use OCP\IAppConfig;
+
+class Service {
```
Comment (PVince81) [inline lib/Marketplace/Service.php]: What is this class and how is it different from `MarketService` ? (add phpdoc)

File: lib/Marketplace/Service.php [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+
+	public function installApp($appId) {
+
+		$info = $this->getInstalledAppInfo($appId);
+		if (!is_null($info)) {
+			throw new \InvalidArgumentException('App is already installed');
+		}
+
+		// download package
+		$package = $this->downloadPackage($appId);
```
Comment (PVince81) [inline lib/Marketplace/Service.php]: I have a sensation of deja vu. Is that the other class for the old API and this one for the market API ?

File: lib/Marketplace/Service.php [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+
+		return $path;
+	}
+
+	public function isAppInstalled($appId) {
+		$info = $this->getInstalledAppInfo($appId);
+		return !is_null($info);
+	}
+
+	public function updateAvailable($appId) {
```
Comment (PVince81) [inline lib/Marketplace/Service.php]: `getAvailableUpdateVersion()`

File: lib/Marketplace/Service.php [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+			$info = $this->appManager->getAppInfo($app);
+			if (isset($info['id']) && $info['id'] === $appId) {
+				return $info;
+			}
+		}
+
+		return null;
+	}
+
+	public function updateApp($appId) {
```
Comment (PVince81) [inline lib/Marketplace/Service.php]: PHPdocs please

File: lib/Marketplace/Service.php [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+					$appId = $info['id'];
+					$newVersion = $this->updateAvailable($appId);
+					if ($newVersion) {
+						$result[$app] = [
+							'version' => $newVersion,
+							'id' => $appId
+						];
+					}
+				} catch (\InvalidArgumentException $ex) {
+				}
```
Comment (PVince81) [inline lib/Marketplace/Service.php]: add comment why it is ok to ignore

File: lib/Marketplace/Service.php [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+	}
+
+	private function getApps() {
+		$version = $this->getPlatformVersion();
+		list($version,) = $this->normalizeVersions($version, '1.2.3');
+		if (!is_null($this->apps)) {
+			return $this->apps;
+		}
+		// read from cache
+		if (\OC::$server->getMemCacheFactory()->isAvailable()) {
```
Comment (PVince81) [inline lib/Marketplace/Service.php]: use DI for MemcacheFactory ?

File: lib/Marketplace/Service.php (hunk 10 above)
Comment (DeepDiver1975) [inline lib/Marketplace/Service.php]: MarketService is basically the dispatcher to the AppStore/Service or the Marketplay/Service

File: lib/AppStore/Service.php (hunk 2 above)
Comment (DeepDiver1975) [inline lib/AppStore/Service.php]: eeeh-no

**Comment by DeepDiver1975** [comment]
@PVince81 MERGE? THX

**Pull request merged**